orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py

```diff
@@ -1072,6 +1072,55 @@ def run_step(model, input):
         _test_helpers.assert_values_are_close(ort_prediction, pt_prediction)
         _test_helpers.assert_values_are_close(ort_input.grad, pt_input.grad)
 
+@pytest.mark.parametrize("equation", ["s,se->se", "se,sc->sec", "se,se->s", "sec,sm->ecm",
+                                      "sec,ecm->sm", "ks,ksm->sm", "kes,ems->mek", "kes,ksm->ms"])
+def test_gradient_correctness_einsum(equation):
```

Comment:
candidates = ["a", "b", "c", "ab", "ba", "ac", "ca", "bc", "cb", "abc", "acb", "bac", "bca", "cab", "cba"]
@pytest.mark.parametrize("left", candidates)
@pytest.mark.parametrize("right", candidates)
def test(left, right): 
    output_pattern = random_generate_a_valid_output_pattern(left, right)
    pattern = "{},{}->{}".format(left, right, output_pattern)

Reply:
I have tried to read the code, but I don't fully understand the fine details, and I feel there are some tricks that I might have missed. 
I feel that the easiest way to gain some confidence on this piece of the code is to have a good coverage on the test. 
I will be comfortably signing off this PR, if we have a test like above.

Reply:
I did that work for tf2onnx, here is the list of unit tests I put in place to check it was working: https://github.com/onnx/tensorflow-onnx/blob/master/tests/test_einsum_helper.py. It tests many permutation for equations with two terms. I added cases from numpy unit tests.

Reply:
Added a new test function to run such combinations (with some sampling), which will take ~1.5mins. It's good that the "ab*ac->cb" case expose one bug from Gemm-Transpose fusion. I also fixed it in the latest commit.

Reply:
Great! Test coverage paying off.